// common/config.js
var npath = require('path')
var _ = require('lodash')
var utils = require('./utils')

var config = {
    root: npath.normalize(__dirname + '/../../'),

    devServerPort: 5678,
    preProdServerPort: 5679,
    prodServerPort: 5680
}

_.extend(config, {
    // matriks 命令所需
    parentPath: utils.p(config.root + '/../'),
    
    // 开发相关路径
    seedRelatedPath: utils.p(config.root + '/seed-related/'),
    localPath: utils.p(config.root + '/locals/'),
    
    destPath: utils.p(config.root + '/dest/'),
    tmpPath: utils.p(config.root + '/tmp/')
})

_.extend(config, {
    // matriks 命令所需
    matriksPath: utils.p(config.parentPath + '/_matriks_/'),
    projectPath: utils.p(config.parentPath + '/project/'),    

    // 前端文件夹
    frontendPath: utils.p(config.localPath + '/frontend/'),
    backendPath: utils.p(config.localPath + '/backend/'),

    // 入口设置文件夹
    entryPath: utils.p(config.localPath + '/frontend/entries/'),

    // 构建输出文件夹 (开发与生产)
    devDestPath: utils.p(config.destPath + '/dev/'),
    prodDestPath: utils.p(config.destPath + '/prod/')
})

_.extend(config, {
    // matriks 命令所需
    seedPath: utils.p(config.matriksPath + '/seed/'),

    // 开发构建输出
    devPagesPath: utils.p(config.devDestPath + '/pages/'),

    // 生产构建输出
    prodPagesPath: utils.p(config.prodDestPath + '/pages/'),
    prodDistPath: utils.p(config.prodDestPath + '/dist/')
})

utils.ensurePath(config.devPagesPath)
utils.ensurePath(config.prodPagesPath)

var adapter = require('../../locals/config/adaptors/config-adaptor')
var config = adapter.process(config)

module.exports = config
